dhcp-relay/dhcp_kern_xdp.c

```diff
@@ -49,77 +56,162 @@ struct {
 } client_vlans SEC(".maps");
 
+//static int memcpy_var(void *to, void *from, __u8 len) {
+//	__u8 *t8 = to, *f8 = from;
+//	int i;
+//
+//	if (len > RAI_OPTION_LEN) {
+//		return -1;
+//	}
+//
+//	for (i = 0; i < len; i++) {
+//
+//		if (i > RAI_OPTION_LEN) {
+//			return -1;
+//		}
+//
+//		*t8++ = *f8++;
+//
+//	}
+//
+//
+//	if (i == RAI_OPTION_LEN)
+//		return -1;
+//
+//	return 0;
+//
+//}
+
+int u16_to_ascii(struct sub_option *opt, __u8 offset, __u16 num) {
+
+	if (opt == NULL)
+		return -1;
+
+	if (offset > RAI_OPTION_LEN)
+		return -1;
+
+	if (offset < U16_ASCII_LEN)
+		return -1;
+
+	__u8 i;
+	//#pragma unroll U16_ASCII_LEN
+	for (i = offset - 1; i > 0; i--) {
+
+		if (i == 0)
+			break;
+
+		opt->val[i] = (__u8) (num % 10) + '0';
+		num /= 10;
+		if (num == 0)
+			break;
+
+	}
+
+	if (i > 0) {
+		opt->val[--i] = '.';
+	}
+
+	if (i > RAI_OPTION_LEN)
+		return -1;
+
+	return i;
+
+}
+
+int str_len(struct dev_name *dev) {
+
+	if (dev == NULL)
+		return -1;
+
+	__u8 i = 0;
+	for (i = 0; i < RAI_OPTION_LEN; i++)
+		if (dev->name[i] == 0)
+			break;
+
+	return i;
+
+}
+
+int copy_dev_name(struct sub_option *dest, __u8 offset, struct dev_name *dev) {
+
+	if (dest == NULL)
+		return -1;
+
+	if (dev == NULL)
+		return -1;
+
+	__u8 i;
+
+	/* Copy device name and left-align VLAN part*/
+#pragma unroll RAI_OPTION_LEN
+	for (i = 0; i < RAI_OPTION_LEN; i++) {
+
+		if (i < dev->len) {
+			/* Copy device name */
+			dest->val[i] = dev->name[i];
+		} else if (offset < RAI_OPTION_LEN) {
+			/* Move VLAN part (all bytes from offset and up) */
+			dest->val[i] = dest->val[offset];
+			dest->val[offset++] = 0;
+		}
+
+	}
+
+	return offset;
+}
+
 /* Inserts DHCP option 82 into the received DHCP packet
  * at the specified offset.
  */
 static __always_inline int write_dhcp_option_82(void *ctx, int offset,
-	struct collect_vlans *vlans, char *dev) {
-	struct dhcp_option_82 option;
+		struct collect_vlans *vlans, struct dev_name dev) {
+
+	struct dhcp_option_82 option = {0};
 
 	option.t = DHO_DHCP_AGENT_OPTIONS;
 	option.len = sizeof (struct sub_option) + sizeof (struct sub_option);
 	option.circuit_id.option_id = RAI_CIRCUIT_ID;
-	option.circuit_id.len = sizeof(option.circuit_id.val);
-	
+	option.circuit_id.len = sizeof (option.circuit_id.val);
+	option.remote_id.option_id = RAI_REMOTE_ID;
+	option.remote_id.len = sizeof (option.remote_id.val);
+
 	/* Reconstruct VLAN device name
 	 * Convert VLAN tags to ASCII from right to left, starting with
 	 * inner VLAN tag.
-	 * Device name is 16 characters long and prepended with dash, e.g.:
-	 * ----ens6f0.83.20
-	 * We avoid null bytes to ensure compatibility with DHCP servers that
-	 * interpret null as a string terminator.
+	 * Device name is up to 16 characters long - remaining buffer space
+	 * contains null bytes.
 	 */
-	 
-	char buf[IF_NAMESIZE];
-	memset(buf, '-', sizeof (buf));
-	
-	int c = VLAN_ASCII_MAX;		/* We will need 4 bytes at most */
-	int i = IF_NAMESIZE - 1;
+
+	int i = RAI_OPTION_LEN;
```

Comment:
The kernel style guidelines forbid mixing variable declarations and code, so please move the declarations to the top of the function. Also, `i` is a terrible variable name (only use that for idiomatic loops), maybe `vlan_offset`? Finally, you can use `sizeof(buf.name)` (and changing the `u16_to_ascii` calls to this `struct dev_name buf` as mentioned above.